### PR TITLE
dispatch: fix missing receivers in Groups()

### LINKED ISSUE #1959: Not all receivers present in /api/v2/alerts/groups responses
**What did you do?**

Query /api/v2/alerts/groups responses continuesly

**What did you expect to see?**

Based on my configuration alerts should go to all receivers, so groups should be generated for each receiver as well. Alerts are tagged with all configured receivers but groups for some receivers are missing.

**What did you see instead? Under which circumstances?**

With the configuration as below:

<details>
  <summary>Configuration</summary>

```YAML
global:
  resolve_timeout: 30s
route:
  group_by: ["alertname"]
  group_wait: 5s
  group_interval: 10s
  repeat_interval: 999h
  receiver: "default"
  routes:
    - receiver: "default"
      group_by: []
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-cluster-service"
      group_by: ["alertname", "cluster", "service"]
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-name"
      group_by: [alertname]
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-cluster"
      group_by: [cluster]
      match_re:
        alertname: .*
      continue: true

inhibit_rules:
  - source_match:
      severity: "critical"
    target_match:
      severity: "warning"
    # Apply inhibition if the alertname and cluster is the same in both
    equal: ["alertname", "cluster"]

receivers:
  - name: "default"
  - name: "by-cluster-service"
  - name: "by-name"
  - name: "by-cluster"
```
</details>

When I query api/v2/alerts/groups I will get alert groups for either `default` or `by-name` receiver, but never both. Responses keep switching between one or the other.
See captured respones with only Receiver field from the group object and from alert objects.
[1.txt](https://github.com/prometheus/alertmanager/files/3381740/1.txt)
[2.txt](https://github.com/prometheus/alertmanager/files/3381741/2.txt)
One shows default receiver groups present, the other by-name receiver groups.

**Environment**

* System information:

```Linux 5.1.16-300.fc30.x86_64 x86_64```

* Alertmanager version:

```
Version Information
Branch: HEAD
BuildDate: 20190708-14:31:49
BuildUser: root@868685ed3ed0
GoVersion: go1.12.6
Revision: 1ace0f76b7101cccc149d7298022df36039858ca
Version: 0.18.0
```

Same behaviour observed with 0.17.0

* Prometheus version:

N/A

* Alertmanager configuration file:

<details>
  <summary>Configuration</summary>

```YAML
global:
  resolve_timeout: 30s
route:
  group_by: ["alertname"]
  group_wait: 5s
  group_interval: 10s
  repeat_interval: 999h
  receiver: "default"
  routes:
    - receiver: "default"
      group_by: []
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-cluster-service"
      group_by: ["alertname", "cluster", "service"]
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-name"
      group_by: [alertname]
      match_re:
        alertname: .*
      continue: true
    - receiver: "by-cluster"
      group_by: [cluster]
      match_re:
        alertname: .*
      continue: true

inhibit_rules:
  - source_match:
      severity: "critical"
    target_match:
      severity: "warning"
    # Apply inhibition if the alertname and cluster is the same in both
    equal: ["alertname", "cluster"]

receivers:
  - name: "default"
  - name: "by-cluster-service"
  - name: "by-name"
  - name: "by-cluster"
```
</details>

* Logs:
Didn't find any logs that would show API response information relevant to this

### PR DESCRIPTION
**Pull request by simonpasquier**
Closes #1959

### CODE REVIEW

File: dispatch/dispatch.go [hunk 1]
```diff
@@ -207,8 +204,13 @@ func (d *Dispatcher) Groups(routeFilter func(*Route) bool, alertFilter func(*typ
 			groups = append(groups, alertGroup)
 		}
 	}
-
 	sort.Sort(groups)
+	for i := range groups {
+		sort.Sort(groups[i].Alerts)
```
Comment (stuartnelson3) [inline dispatch/dispatch.go]: great idea

Comment (simonpasquier) [reply dispatch/dispatch.go]: It was very much needed for the tests to pass :)

**Comment by simonpasquier** [comment]
@prymitive can you tell me if it fixes the issue for you?

**Comment by prymitive** [comment]
I can do some testing after work and update here

**Comment by prymitive** [comment]
Test using master branch:
===================

```
$ ./alertmanager --version
alertmanager, version 0.18.0 (branch: HEAD, revision: f450720213de12237437552fa8e86b715e0baafa)
  build user:       lukasz@kwakBook
  build date:       20190723-21:56:49
  go version:       go1.12.7
```

```
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  15 {"name":"by-cluster-service"}
  24 {"name":"by-name"}
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  15 {"name":"by-cluster-service"}
  24 {"name":"by-name"}
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  15 {"name":"by-cluster-service"}
  24 {"name":"default"}
```

There's either `default` or `by-name` receiver, but never both, so we hit this issue.

Test using this PR as HEAD:
====================

```
$ ./alertmanager --version
alertmanager, version 0.18.0 (branch: pr/1964, revision: 110dd7555c254a3409515d1d7289e6833b7acd1a)
  build user:       lukasz@kwakBook
  build date:       20190723-21:45:37
  go version:       go1.12.7
```

```
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  16 {"name":"by-cluster-service"}
  13 {"name":"by-name"}
  13 {"name":"default"}
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  15 {"name":"by-cluster-service"}
  12 {"name":"by-name"}
  12 {"name":"default"}
kwakBook:alertmanager lukasz$ curl -s localhost:9093/api/v2/alerts/groups | jq -c '.[] | .receiver' | sort | uniq -c
   3 {"name":"by-cluster"}
  16 {"name":"by-cluster-service"}
  13 {"name":"by-name"}
  13 {"name":"default"}
```

Both receivers are always present as expected, so the issue is gone.

**Comment by fcastello** [comment]
Can it be that this issue was not fully fixed?
I am still seeing duplicates when I add multiple receivers
`    - match:
        destination: webhook
      receiver: webhook
      continue: true
    - match:
        destination: webhook
      receiver: slack
`
That is the part of my config that generates the duplicate alerts in the UI, Also they are inheriting the group_by from the main config.
I am running the image with master tag from dockerhub which states that was built 2 days ago so this fix should be in there

**Comment by simonpasquier** [comment]
@fcastello can you share a screenshot and the full AlertManager configuration? I tried to reproduce but couldn't spot the issue.

**Pull request merged**